shared/domains/chelonia/internals.js

```diff
@@ -216,22 +401,54 @@ export default (sbp('sbp/selectors/register', {
     if (config.preOp) {
       processOp = config.preOp(message, state) !== false && processOp
     }
+
+    // Signature verification
+    // TODO: Temporary. Skip verifying default signatures
+    if (!isNaN(0) && signature.type !== 'default') {
+      // This sync code has potential issues
+      // The first issue is that it can deadlock if there are circular references
+      // The second issue is that it doesn't handle key rotation. If the key used for signing is invalidated / removed from the originating contract, we won't have it in the state
+      // Both of these issues can be resolved by introducing a parameter with the message ID the state is based on. This requires implementing a separate, ephemeral, state container for operations that refer to a different contract.
+      // The difficulty of this is how to securely determine the message ID to use.
+      // The server can assist with this.
+
+      const authorizedKeys = opT === GIMessage.OP_CONTRACT ? keysToMap(((opV: any): GIOpContract).keys) : state._vm.authorizedKeys
+      let signingKey = authorizedKeys?.[signature.keyId]
+
+      if (!signingKey && opT !== GIMessage.OP_CONTRACT && message.originatingContractID() !== message.contractID()) {
+        const originatingContractState = await sbp('chelonia/withEnv', message.originatingContractID(), { skipActionProcessing: true }, [
+          'chelonia/latestContractState', message.originatingContractID()
+        ])
+
+        signingKey = originatingContractState._vm?.authorizedKeys?.[signature.keyId]
+      }
```

Comment:
What scenario is this `if` condition checking for, and what is going on here?

Reply:
`signingKey` may not be present in the contract. This happens in cross-contract interactions, where a contract writes to another contract using its own keys. For example, when one requests to join a group, that message cannot be signed by the group because the secret key is only known to group members. Instead, it is signed with the keys of the person joining.